[Script Tracking Privacy] Add support for per-category, per-domain script access rules

whsieh opened the pull request:
#### 3f3a7da0797509e831f0348ff46469d186c5dc7a
<pre>
[Script Tracking Privacy] Add support for per-category, per-domain script access rules
<a href="https://bugs.webkit.org/show_bug.cgi?id=296091">https://bugs.webkit.org/show_bug.cgi?id=296091</a>
<a href="https://rdar.apple.com/156004348">rdar://156004348</a>

Reviewed by Abrar Rahman Protyasha.

Add support for per-domain script access rules to allow script access by category. This makes it
possible to deploy granular filtering rules that would (for instance) allow a specific known
fingerprinting or tracking script to programmatically access text field values or query parameters
to mitigate compatibility issues, but keep the remaining script access categories locked down with
entropy reduction (i.e. noise injection, or returning fixed values).

See below for more details.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::requiresScriptTrackingPrivacyProtection):

Implement the main fix here — return early with `false` in the case where the script access category
is allowed for the given tainted script URL, by calling into the new helper method on `Page` below.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::shouldAllowScriptAccess const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::shouldAllowScriptAccess const):

Add plumbing through `ChromeClient`, to return whether or not script access for a known tainted URL
should be allowed.

* Source/WebCore/page/Page.h:
* Source/WebCore/page/ScriptTrackingPrivacyCategory.cpp:
(WebCore::description):
(WebCore::scriptCategoryAsFlag):
* Source/WebCore/page/ScriptTrackingPrivacyCategory.h:

Introduce `ScriptTrackingPrivacyFlag`, which represents each of the `ScriptTrackingPrivacyCategory`
enum values, but bit-shifted in a way that makes them suitable for use in an `OptionSet`. We also
remove the currently-unused `Unspecified` value, which ensures that the first category enum value
now starts at 0 instead of 1, leaving no holes in the corresponding flags.

* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::getAllowedCategories):

Map from `WPScriptAccessCategories` to `OptionSet&lt;ScriptTrackingPrivacyFlag&gt;`.

(WebKit::ScriptTrackingPrivacyController::updateList):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/ScriptTrackingPrivacyFilter.cpp:

Change how `ScriptTrackingPrivacyFilter` works. Instead of keeping track of multiple hash sets of
domain names, we turn this into a `HashMap` of host name to `OptionSet&lt;ScriptTrackingPrivacyFlag&gt;`,
representing the allowed script access categories for each host. We also keep track of all
categories that don&apos;t appear in *any* allow-list of categories, for any host in the list; we can
exit early and avoid the (slightly more costly) hash lookup for these categories, if the category in
question is not in `m_categoriesWithAllowedHosts`.

(WebKit::initializeFilterRules):
(WebKit::ScriptTrackingPrivacyFilter::ScriptTrackingPrivacyFilter):
(WebKit::ScriptTrackingPrivacyFilter::lookup):

Refactor `matches()` into a private helper, `lookup()`, that returns whether or not the given URL
was a match, as well as the set of allowed categories. Reimplement `matches()` in terms of this new
helper, and also use this helper to implement the new `shouldAllowAccess` method below.

(WebKit::ScriptTrackingPrivacyFilter::matches):
(WebKit::ScriptTrackingPrivacyFilter::shouldAllowAccess):
* Source/WebKit/Shared/ScriptTrackingPrivacyFilter.h:
* Source/WebKit/Shared/ScriptTrackingPrivacyFilter.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::shouldAllowScriptAccess const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::requiresScriptTrackingPrivacyProtections const):
(WebKit::WebProcess::shouldAllowScriptAccess const):
* Source/WebKit/WebProcess/WebProcess.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScriptTrackingPrivacyTests.mm:
(-[TestWPFingerprintingScript initWithHost:isFirstParty:isTopDomain:allowedCategories:]):
(TestWebKitAPI::makeFingerprintingScriptsRequestHandler):
(TestWebKitAPI::FingerprintingScriptsRequestSwizzler::FingerprintingScriptsRequestSwizzler):
(TestWebKitAPI::(ScriptTrackingPrivacyTests, ScriptAccessCategories)):
(-[TestWPFingerprintingScript initWithHost:isFirstParty:isTopDomain:]): Deleted.

Add an API test to exercise this mechanism, by allow-listing form control access in the (simulated)
fingerprinter scripts dataset and verifying that the tainted script can still read form control
values, but not speech synthesis voices (the latter of which is not in the allowed set).

Canonical link: <a href="https://commits.webkit.org/297519@main">https://commits.webkit.org/297519@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a28ed8bfd7c22f0f21f5df53e4e8f817814525bd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112038 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31709 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118061 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62277 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/114000 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32385 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40286 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85112 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35784 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114985 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25865 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100810 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65545 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/111470 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25179 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18950 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61913 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95258 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19025 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121380 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39071 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93958 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39452 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97066 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93777 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23937 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38981 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16769 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35093 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38965 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44477 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38602 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41930 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40318 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/e04ac4156bab3546a9f4039d414604e8989bca1f. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48157

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5e92244cdd57c3ebfcd0a286bbd8f5f67730eb5b. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48157

whsieh commented on Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
```diff
@@ -751,6 +751,38 @@ void configureForAdvancedPrivacyProtections(NSURLSession *) { }
 
+#if ENABLE(SCRIPT_TRACKING_PRIVACY_PROTECTIONS)
+
+static WebCore::ScriptTrackingPrivacyFlags getAllowedCategories(WPScriptAccessCategories categories)
```
Good point — `allowedScriptTrackingCategories` should work!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a28ed8bfd7c22f0f21f5df53e4e8f817814525bd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/48157

whsieh committed:
[Script Tracking Privacy] Add support for per-category, per-domain script access rules
https://bugs.webkit.org/show_bug.cgi?id=296091
rdar://156004348

Reviewed by Abrar Rahman Protyasha.

Add support for per-domain script access rules to allow script access by category. This makes it
possible to deploy granular filtering rules that would (for instance) allow a specific known
fingerprinting or tracking script to programmatically access text field values or query parameters
to mitigate compatibility issues, but keep the remaining script access categories locked down with
entropy reduction (i.e. noise injection, or returning fixed values).

See below for more details.

* Source/WebCore/PAL/pal/spi/cocoa/WebPrivacySPI.h:
* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::requiresScriptTrackingPrivacyProtection):

Implement the main fix here — return early with `false` in the case where the script access category
is allowed for the given tainted script URL, by calling into the new helper method on `Page` below.

* Source/WebCore/page/ChromeClient.h:
(WebCore::ChromeClient::shouldAllowScriptAccess const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::shouldAllowScriptAccess const):

Add plumbing through `ChromeClient`, to return whether or not script access for a known tainted URL
should be allowed.

* Source/WebCore/page/Page.h:
* Source/WebCore/page/ScriptTrackingPrivacyCategory.cpp:
(WebCore::description):
(WebCore::scriptCategoryAsFlag):
* Source/WebCore/page/ScriptTrackingPrivacyCategory.h:

Introduce `ScriptTrackingPrivacyFlag`, which represents each of the `ScriptTrackingPrivacyCategory`
enum values, but bit-shifted in a way that makes them suitable for use in an `OptionSet`. We also
remove the currently-unused `Unspecified` value, which ensures that the first category enum value
now starts at 0 instead of 1, leaving no holes in the corresponding flags.

* Source/WebKit/Platform/cocoa/WebPrivacyHelpers.mm:
(WebKit::getAllowedCategories):

Map from `WPScriptAccessCategories` to `OptionSet<ScriptTrackingPrivacyFlag>`.

(WebKit::ScriptTrackingPrivacyController::updateList):
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/ScriptTrackingPrivacyFilter.cpp:

Change how `ScriptTrackingPrivacyFilter` works. Instead of keeping track of multiple hash sets of
domain names, we turn this into a `HashMap` of host name to `OptionSet<ScriptTrackingPrivacyFlag>`,
representing the allowed script access categories for each host. We also keep track of all
categories that don't appear in *any* allow-list of categories, for any host in the list; we can
exit early and avoid the (slightly more costly) hash lookup for these categories, if the category in
question is not in `m_categoriesWithAllowedHosts`.

(WebKit::initializeFilterRules):
(WebKit::ScriptTrackingPrivacyFilter::ScriptTrackingPrivacyFilter):
(WebKit::ScriptTrackingPrivacyFilter::lookup):

Refactor `matches()` into a private helper, `lookup()`, that returns whether or not the given URL
was a match, as well as the set of allowed categories. Reimplement `matches()` in terms of this new
helper, and also use this helper to implement the new `shouldAllowAccess` method below.

(WebKit::ScriptTrackingPrivacyFilter::matches):
(WebKit::ScriptTrackingPrivacyFilter::shouldAllowAccess):
* Source/WebKit/Shared/ScriptTrackingPrivacyFilter.h:
* Source/WebKit/Shared/ScriptTrackingPrivacyFilter.serialization.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::shouldAllowScriptAccess const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::requiresScriptTrackingPrivacyProtections const):
(WebKit::WebProcess::shouldAllowScriptAccess const):
* Source/WebKit/WebProcess/WebProcess.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ScriptTrackingPrivacyTests.mm:
(-[TestWPFingerprintingScript initWithHost:isFirstParty:isTopDomain:allowedCategories:]):
(TestWebKitAPI::makeFingerprintingScriptsRequestHandler):
(TestWebKitAPI::FingerprintingScriptsRequestSwizzler::FingerprintingScriptsRequestSwizzler):
(TestWebKitAPI::(ScriptTrackingPrivacyTests, ScriptAccessCategories)):
(-[TestWPFingerprintingScript initWithHost:isFirstParty:isTopDomain:]): Deleted.

Add an API test to exercise this mechanism, by allow-listing form control access in the (simulated)
fingerprinter scripts dataset and verifying that the tainted script can still read form control
values, but not speech synthesis voices (the latter of which is not in the allowed set).

Canonical link: https://commits.webkit.org/297519@main
